Legacy Expression: Remove next reference

georgwiese opened the pull request:
Builds on #2890

This PR removes the "next" flag from `AlgebraicReference`. This requires a few changes, because OpenVM's AIRs do use them. It should still be removed, because we can't deal with it when generating APCs.

The solution is to introduce a new type that wraps `AlgebraicReference`:
https://github.com/powdr-labs/powdr/blob/1cee98dbba30051ce915566175499f8c28ee831b/openvm/src/utils.rs#L25-L30

Note that this also takes out the `is_{first,last,transition}` reference, which we also can't handle when compiling APCs!

When exporting OpenVM's chips to PIL, we use the `OpenVmReference` type; for everything else, we convert to `AlgebraicReference`. For this, we need to make sure that we don't do that for blacklisted opcodes (because they might contain next references).

georgwiese commented on openvm/src/utils.rs:
```diff
@@ -108,49 +139,34 @@ pub fn symbolic_to_algebraic<T: PrimeField32, P: FieldElement>(
-        SymbolicExpression::IsFirstRow => AlgebraicExpression::Reference(AlgebraicReference {
-            name: "is_first_row".to_string(),
-            poly_id: PolyID {
-                id: 0,
```
Looks like we used to have ID collisions before this PR!

pacheco commented on openvm/src/lib.rs:
```diff
@@ -636,21 +641,27 @@ where
                     let constraints = get_constraints(air);
 
+                    // Note that the unwrap() calls on the results of try_convert and openvm_bus_interaction_to_powdr
+                    // are safe as long as `used_instructions` does not contain any opcodes that are not supported
```
should be this comment be moved to the function doc?

georgwiese commented:
I rebased to `main` and refactored a bit in 022ce4efc8dd098323ad0b0e4da4c28f3ae46a63. I think the refactoring should make things clearer and fix both of your comments :)

georgwiese commented on openvm/src/instruction_blacklist.rs:
```diff
@@ -0,0 +1,81 @@
+    [
+        Rv32HintStoreOpcode::HINT_STOREW.global_opcode().as_usize(), // contain next references that don't work with apc
+        Rv32HintStoreOpcode::HINT_BUFFER.global_opcode().as_usize(), // contain next references that don't work with apc
+        Rv32LoadStoreOpcode::LOADB.global_opcode().as_usize(),
```
@leonardoalt Can you remind me why we blacklist loadb and loadh?

Schaeff commented on openvm/src/utils.rs:
```diff
@@ -13,17 +14,53 @@ use openvm_stark_backend::{
 };
 use powdr_number::{BabyBearField, FieldElement};
 
+pub enum OpenVmReference {
+    /// Reference to a witness column. The boolean indicates if the reference is to the next row.
+    WitnessColumn(AlgebraicReference, bool),
+    IsFirstRow,
+    IsLastRow,
+    IsTransition,
+}
```
I find it odd that we can't we access `first_row'`. Shouldn't columns and offsets be orthogonal?

I had a proposal ages ago which here would look like:
```
enum Expression {
   Reference(ShiftedReference)
}

struct ShiftedReference {
   reference: Reference,
   shift: bool // or an number for larger offsets
}

enum Reference {
   FirstRow,
   Witness(WitnessReference)
}
```
